File: java-automation/src/test/java/com/myecommerce/automation/dsl/steps/ProductDetailSteps.java
package com.myecommerce.automation.dsl.steps;

import com.myecommerce.automation.dsl.protocols.CartProtocol;
import com.myecommerce.automation.dsl.protocols.CatalogueProtocol;
import com.myecommerce.automation.dsl.protocols.DriverFactory;
import io.cucumber.java.en.Given;
import io.cucumber.java.en.Then;
import io.cucumber.java.en.When;
import lombok.extern.java.Log;
import org.assertj.core.api.SoftAssertions;

import java.util.function.Supplier;

import static org.assertj.core.api.Assertions.assertThat;

@Log
public class ProductDetailSteps {

    private final CatalogueProtocol catalogue = DriverFactory.createCatalogue();
    private final Supplier<CartProtocol> cart = DriverFactory::createCart;
    private String capturedListingTitle;
    private int    notedCartCount;

    @Given("the shopper is viewing the first product")
    public void shopperIsViewingFirstProduct() {
        catalogue.browseCatalogue();
        capturedListingTitle = catalogue.getProductListing().cards().getFirst().title();
        catalogue.viewFirstProduct();
        log.fine("navigated to first product detail; listing title was: '" + capturedListingTitle + "'");
    }

    @Given("the shopper notes the cart count on the detail page")
    public void shopperNotesCartCountOnDetailPage() {
        notedCartCount = cart.get().getCartState().itemCount();
        log.fine("noted cart count: " + notedCartCount);
    }

    @When("the shopper clicks the first product card")
    public void shopperClicksFirstProductCard() {
        catalogue.viewFirstProduct();
        log.fine("clicked first product card");
    }

    @When("the shopper adds the product to the cart from the detail page")
    public void shopperAddsProductFromDetailPage() {
        cart.get().addProductToCart();
        log.fine("added product to cart from detail page");
    }

    @When("the shopper navigates back")
    public void shopperNavigatesBack() {
        catalogue.returnToProductListing();
        log.fine("navigated back to product listing");
    }

    @Then("the URL should match the product detail pattern")
    public void urlShouldMatchProductDetailPattern() {
        assertThat(catalogue.currentUrl())
            .as("URL should match product detail pattern")
            .matches(".*/products/\\d+");
        log.fine("URL matches product detail pattern");
    }

    @Then("the product title should be visible on the detail page")
    public void productTitleShouldBeVisibleOnDetailPage() {
        assertThat(catalogue.getProductDetail().title())
            .as("product title should be visible and non-blank on detail page")
            .isNotBlank();
        log.fine("product title is visible on detail page");
    }

    @Then("the product detail page should show price, description and image")
    public void detailPageShouldShowPriceDescriptionAndImage() {
        var detail = catalogue.getProductDetail();
        SoftAssertions soft = new SoftAssertions();
        soft.assertThat(detail.price())
            .as("price should show a $ value")
            .matches(".*\\$\\d+.*");
        soft.assertThat(detail.description())
            .as("description should not be blank")
            .isNotBlank();
        soft.assertThat(detail.imagePresent())
            .as("product image should be present")
            .isTrue();
        soft.assertAll();
        log.fine("detail page shows price, description, and image");
    }

    @Then("the product title should match the one from the listing")
    public void productTitleShouldMatchTheListing() {
        assertThat(catalogue.getProductDetail().title())
            .as("detail title should contain the listing title")
            .contains(capturedListingTitle.trim());
        log.fine("detail title matches listing title: '" + capturedListingTitle + "'");
    }

    @Then("the add to cart button should show Added to Cart and be disabled")
    public void addToCartButtonShouldShowAddedToCartAndBeDisabled() {
        var detail = catalogue.getProductDetail();
        SoftAssertions soft = new SoftAssertions();
        soft.assertThat(detail.addToCartButtonText())
            .as("add-to-cart button text should say 'Added to Cart'")
            .containsIgnoringCase("Added to Cart");
        soft.assertThat(detail.addToCartEnabled())
            .as("add-to-cart button should be disabled after adding")
            .isFalse();
        soft.assertAll();
        log.fine("add-to-cart button shows 'Added to Cart' and is disabled");
    }

    @Then("the shopper should be back on the product listing")
    public void shopperShouldBeBackOnProductListing() {
        SoftAssertions soft = new SoftAssertions();
        soft.assertThat(catalogue.getProductListing().cards())
            .as("product listing should be visible after navigating back")
            .isNotEmpty();
        soft.assertThat(catalogue.currentUrl())
            .as("URL should be back on the listing")
            .matches(".*/(\\?.*)?$");
        soft.assertAll();
        log.fine("shopper is back on product listing");
    }
}

File: kotlin-automation/src/test/kotlin/com/myecommerce/automation/dsl/steps/ProductDetailSteps.kt
package com.myecommerce.automation.dsl.steps

import com.myecommerce.automation.dsl.protocols.createCart
import com.myecommerce.automation.dsl.protocols.createCatalogue
import io.cucumber.java.en.Given
import io.cucumber.java.en.Then
import io.cucumber.java.en.When
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.SoftAssertions

class ProductDetailSteps {

    private val catalogue = createCatalogue()
    private val cart by lazy { createCart() }

    private var capturedListingTitle: String = ""
    private var notedCartCount: Int = 0

    @Given("the shopper is viewing the first product")
    fun shopperIsViewingFirstProduct() {
        catalogue.browseCatalogue()
        capturedListingTitle = catalogue.getProductListing().cards.first().title
        catalogue.viewFirstProduct()
    }

    @Given("the shopper notes the cart count on the detail page")
    fun shopperNotesCartCountOnDetailPage() {
        notedCartCount = cart.getCartState().itemCount
    }

    @When("the shopper clicks the first product card")
    fun shopperClicksFirstProductCard() = catalogue.viewFirstProduct()

    @When("the shopper adds the product to the cart from the detail page")
    fun shopperAddsProductFromDetailPage() = cart.addProductToCart()

    @When("the shopper navigates back")
    fun shopperNavigatesBack() = catalogue.returnToProductListing()

    @Then("the URL should match the product detail pattern")
    fun urlShouldMatchProductDetailPattern() {
        assertThat(catalogue.currentUrl())
            .`as`("URL should match product detail pattern")
            .matches(".*/products/\\d+")
    }

    @Then("the product title should be visible on the detail page")
    fun productTitleShouldBeVisibleOnDetailPage() {
        assertThat(catalogue.getProductDetail().title)
            .`as`("product title should be visible and non-blank on detail page")
            .isNotBlank
    }

    @Then("the product detail page should show price, description and image")
    fun detailPageShouldShowPriceDescriptionAndImage() {
        val detail = catalogue.getProductDetail()
        SoftAssertions().apply {
            assertThat(detail.price).`as`("price should show a \$ value").matches(".*\\$\\d+.*")
            assertThat(detail.description).`as`("description should not be blank").isNotBlank
            assertThat(detail.imagePresent).`as`("product image should be present").isTrue
            assertAll()
        }
    }

    @Then("the product title should match the one from the listing")
    fun productTitleShouldMatchTheListing() {
        assertThat(catalogue.getProductDetail().title)
            .`as`("detail title should contain the listing title")
            .contains(capturedListingTitle.trim())
    }

    @Then("the add to cart button should show Added to Cart and be disabled")
    fun addToCartButtonShouldShowAddedToCartAndBeDisabled() {
        val detail = catalogue.getProductDetail()
        SoftAssertions().apply {
            assertThat(detail.addToCartButtonText)
                .`as`("add-to-cart button text should say 'Added to Cart'")
                .containsIgnoringCase("Added to Cart")
            assertThat(detail.addToCartEnabled)
                .`as`("add-to-cart button should be disabled after adding")
                .isFalse
            assertAll()
        }
    }

    @Then("the shopper should be back on the product listing")
    fun shopperShouldBeBackOnProductListing() {
        SoftAssertions().apply {
            assertThat(catalogue.getProductListing().cards)
                .`as`("product listing should be visible after navigating back")
                .isNotEmpty
            assertThat(catalogue.currentUrl())
                .`as`("URL should be back on the listing")
                .matches(".*/(\\?.*)?$")
            assertAll()
        }
    }
}
